orc8r/gateway/python/magma/ctraced/trace_manager.py

```diff
@@ -108,39 +120,112 @@ def end_trace(self) -> EndTraceResult:
             stopped = self._stop_trace()
             if not stopped:
                 return EndTraceResult(False, None)
-            while True:
-                if self._ensure_trace_file_exists():
-                    logging.info("TraceManager: Trace file written!")
-                    break
-                logging.info("TraceManager: Waiting 1s for trace file to be "
-                             "written...")
-                time.sleep(1)
+            self._wait_til_trace_file_exists()
 
         # Perform postprocessing of capture file with TShark display filters
-        if len(self._display_filters) > 0:
-            succeeded = self._postprocess_trace()
-            if not succeeded:
-                return EndTraceResult(False, None)
-
-            # Read trace data into bytes
-            with open(self._trace_filename_postprocessed, "rb") as trace_file:
-                data = trace_file.read()  # type: bytes
-        else:
-            # Read trace data into bytes
-            with open(self._trace_filename, "rb") as trace_file:
-                data = trace_file.read()  # type: bytes
+        succeeded = self._conditionally_postprocess_trace()
+        if not succeeded:
+            return EndTraceResult(False, None)
 
-        # Ensure the tmp trace file is deleted
-        self._ensure_tmp_file_deleted()
-        self._trace_filename = ""
-
-        self._is_active = False
+        data = self._get_final_trace_data() # type: bytes
 
+        self._cleanup_trace()
         logging.info("TraceManager: Call trace has ended")
 
         # Everything cleaned up, return bytes
         return EndTraceResult(True, data)
 
+    def _execute_start_trace_command(self, command: List[str]) -> bool:
+        """Executes a command to start a call trace
+
+        Args:
+            command: Shell command with each token ordered in the list.
+              example: ["tshark", "-i", "eth0"] would be for "tshark -i eth0"
+
+        Returns:
+            True if successfully executed command
+        """
+        logging.info("TraceManager: Starting trace with %s, command: [%s]",
+                     self._tool_name, ' '.join(command))
+
+        self._ensure_trace_directory_exists()
+
+        # TODO(andreilee): Handle edge case where only one instance of the
```

Comment:
What's this case ?

Reply:
somebody could hypothetically start tshark on the AGW, and then calling ctraced through grpc would cause issues.